File: freecodecamp.js
function meetBonfire(argument) {
  // Good luck!
  console.log("you can read this function's argument in the developer tools", argument);

  return true;
}

meetBonfire("You can do this!");

function reverseString(str) {
  var splitString = str.split("");
  var reverseString = splitString.reverse();
  var joinedString = reverseString.join("");
  return joinedString;
}

reverseString("hello");

//INPUT: String
//Convert the string into lowercase
//Strip out all non abc characters
// store that word in a variable
//Then take the string place it in a array, reverse it , convert it back to a string
//Store this reversed string into reversed string variable
//Compare the forward and reverse string
//OUTPUT: return true if it's palindrome. False otherwise
// function palindrome(str) {
//     var lowerCaseString = str.toLowerCase();
//     var splitString = lowerCaseString.split(" ").join("");
//     var reversedString = lowerCaseString.split("").reverse();
//     var joinString = reversedString.join("");
//     console.log(forwardString);
//     console.log(joinString);

//     if (forwardString === joinString) {
//         console.log(true);
//     } else {
//         console.log(false);
//     };

// };

// palindrome("Eye");
// palindrome("race car");
// palindrome("nope");
// palindrome("A man, a plan, a canal. Panama");

// input string
// transform each word to lowercase
// if the letter is at char at 0
// then cap it
// else return full letter
// create a empty arry and push each letter into it
// output every word is CAP

// function titleCase(str) {
//   var titleCaseWord = [];
//   var lowerCase = str.toLowerCase().split(" ");
//   for (var i = 0; i <= lowerCase.length; i++) {
//      lowerCase[i].charAt(0).toUpperCase() + lowercase[i];
//        console.log(lowerCase);
//   }
// }

// titleCase("I'm a little tea pot");

// INPUT: Obtain a sentence
// STEPS:
// Split the sentence into individual words
// Store the results in the variable wordsArray
// Create a variable called maxWordLength and set that equal to 0
// FOR each word in the variable THEN
//     Find the length of each word
//     Compare the length of that word to current value in maxWordLength
// IFthe length is greater than maxWordLength THEN
//     set maxWordLength = to word length
// END IF
// END FOR
// OUTPUT: Return the length of the longest word in the sentence


function findLongestWord(str) {
  var wordArray = str.split(" ");
  var maxWordLength = 0;

  for (var i = 0; i < wordArray.length; i++)  {
        if (wordArray[i].length > maxWordLength)
          maxWordLength = wordArray[i].length;
  }

  console.log(maxWordLength);

};

findLongestWord("The quick brown fox jumped over the lazy dog");
findLongestWord("May the force be with you");
findLongestWord("Google do a barrel roll");
findLongestWord("What is the average airspeed velocity of an unladen swallow");
findLongestWord("What if we try a super-long word such as otorhinolaryngology");

// INPUT: Obtain a array with 4 sub arrays inside
// STEPS:
// Create a variable called maxSubArrayValues and set that equal to []
// Create a variable called maxValue and set it to 0
// FOR each sub array  THEN
//    loop through each value in each sub array
//     Compare the value to the value in the maxValue
// IF the value is greater than one in maxValue THEN
//     push that value into maxSubArrayValues
// END IF
// END FOR
// OUTPUT: Return an array with the max value of each sub array

function largestOfFour(array) {
    var maxSubArrayValues = [];
    var maxValue = 0;

    array.map(function(obj) {
       maxSubArrayValues.push(Math.max.apply(null, obj));
    });
    console.log(maxSubArrayValues);
}

largestOfFour([[13, 27, 18, 26], [4, 5, 1, 3], [32, 35, 37, 39], [1000, 1001, 857, 1]])
largestOfFour([[4, 9, 1, 3], [13, 35, 18, 26], [32, 35, 97, 39], [1000000, 1001, 857, 1]])

// INPUT: Obtain a string and a number
// STEPS:
// store original string into variable called wordToAddArray
// IF number is < 1THEN
//     return an empty string
// ELSE IF number === 1 THEN
//    return the original string
// ELSE
//    FOR (While counter is less than string) THEN
//        push string into wordToAddARRy
//    END FOR
// END IF
// OUTPUT: Return the string repeated n amount of times

function repeat(str, num ) {
  var strArray = str.split(" ");
  var joinArray;

    if (num < 1) {
      return " " ;
    } else if (num === 1) {
      return str;
    } else {
        for (var i = 0; i < (num - 1); i++) {
            strArray.push(str);
        }
    }
    return strArray.join("");
}

repeat("*", 3);
repeat("abc", 3);
repeat("abc", 4);
repeat("abc", 1);
repeat("*", 8);
repeat("abc", -2);

// INPUT: Obtain a array and a number
// STEPS:
// Create a empty container object called split array and set it eql to splitArray
// Iterate through the array n amount of times
//    Push each number into the new array
// ELSE
//    return the original array
// OUTPUT: Return the string split into subarrays based on the number in the argument

function chunk(arr, size) {
    var splitArray = [];

    for (var i = 0; i < arr.length ; i++) {
        if (i >= size)>
    }
}

console.log(chunk(["a", "b", "c", "d"], 2));
console.log(chunk([0, 1, 2, 3, 4, 5], 3));
console.log(chunk([0, 1, 2, 3, 4, 5], 2));
console.log(chunk([0, 1, 2, 3, 4, 5], 4));

function Dog(name, breed, weight) {
    this.name =  name;
    this.breed = breed;
    this.weight = weight;
}

var fido = new Dog("Fido", "Mixed", 38);
var fluffy = new Dog("Fluffy", "Poodle", 30);
var spot = new Dog("Spot", "Chihuahua", 10);
var dogs = [fido, fluffy, spot];

for (var i = 0; i < dogs.length; i++) {
  var size =
}

function Duck(type, canFly) {
    this.type = type;
    this.canFly = true;
}

File: free_code_camp.rb
#Basic algorithm problems
def meetBonfire
    "You can read this text"
    return true
end

#++++++++++++++++++++++++++++++
def reverseString(word)
  reversed_word = word.reverse
  p reversed_word
end

reverseString("hello")
reverseString("Howdy")
reverseString("Greetings from earth")

#++++++++++++++++++++++++++++++
def factorialize(number)

  if number == 0
    p 1
  else
   p (1..number).reduce(:*)
  end

end

factorialize(5)
factorialize(10)
factorialize(20)
factorialize(0)

#++++++++++++++++++++++++++++++
#palindromes
#INPUT: Word
#Change the word to lowercase and strip out all spaces and punctuation
#Reverse the word
#Compare the original word to the reversed word
# IF true THEN
#     return true
# Else
#    return false
#END IF
#OUTPUT: Return true if the given string is spelled the same forward and backwards. Return false if it's not

def palindrome(word)
  word = word.strip.downcase.split("")
  reverse_word = word.reverse

  #word.each_byte do |byte|

#end
end

palindrome(" Eye-ball ")

#++++++++++++++++++++++++++++++
def find_longest_word(sentence)
    word_array = sentence.split(" ")
    max_word = word_array.max_by {|word| word.length }
    p max_word.length
end

find_longest_word("The quick brown fox jumped over the lazy dog.")
find_longest_word("May the force be with you")
find_longest_word("Google do a barrel roll")
find_longest_word("What is the average airspeed velocity of an unladen swallow")
find_longest_word("What if we try a super-long word such as otorhinolaryngology")


#++++++++++++++++++++++++++++++
def title_case(sentence)
    sentence_array = sentence.split(" ")
    sentence_array.map! do |word|
       word.capitalize
    end

    p sentence_array
end
title_case("I'm a little tea pot")
title_case("sHoRt AnD sToUt")
title_case("HERE IS MY HANDLE HERE IS MY SPOUT")

#++++++++++++++++++++++++++++++
def largest_of_four(array)
  max_of_each_row = []
  array.each do |row|
      max_of_each_row << row.max
  end
  p max_of_each_row
end

largest_of_four([[13, 27, 18, 26], [4, 5, 1, 3], [32, 35, 37, 39], [1000, 1001, 857, 1]])
largest_of_four([[4, 9, 1, 3], [13, 35, 18, 26], [32, 35, 97, 39], [1000000, 1001, 857, 1]])

#++++++++++++++++++++++++++++++++
def confirm_ending(string, target)
  last_value = string.split(" ").last
  p last_value
  p last_value.include?(target) ? "True" : "False"
end

confirm_ending("Bastian", "n")
confirm_ending("Connor", "n")
confirm_ending("Walking on water and developing software from a specification are easy if both are frozen", "specification")
confirm_ending("He has to give me a new name", "name")
confirm_ending("He has to give me a new name", "me")
confirm_ending("He has to give me a new name", "na")
confirm_ending("If you want to save our world, you must hurry. We dont know how much longer we can withstand the nothing", "mountain")

#++++++++++++++++++++++++++++++
def repeat(string, number)
  number.times do
    string = string + string
  end
end

repeat("*", 3)
repeat("abc", 3)
repeat("abc", 4)
repeat("abc", 1)
repeat("*", 8)
repeat("abc", -2)

#INPUT: Sentence & Number
#IF sentence.length > number THEN
#   append ... to the end of the sentence
#   full sentence length should now be 11
#IF sentence.length <= 3 THEN
#   return sentence+ ...
#OUTPUT: Return truncated sentence if length > number
def truncate(string,number)
  potential_word = string + "..."
  if potential_word.length > number
      number = number-1
      sliced_string = potential_word.slice(-3..-1)
      sliced_string
      # p truncated_word = sliced_string + "..."
      # p truncated_word.length
      # p sliced_string.length
  end
end
 truncate("A-tisket a-tasket A green and yellow basket", 11)

#++++++++++++++++++++++++++++++
def chunk(array, size)
  array.each_slice(size) { |number| p number}
end

chunk(["a", "b", "c", "d"],2)

#++++++++++++++++++++++++++++++
#INPUT: Obtain a collection of values and number
#Starting from the beginning of the collection
  #Delete values based on the number
  #Return the new collection
#OUTPUT: Return the shrunken collection based on the value of the number

def slasher(array, how_many)
 p array.drop(how_many)
end

slasher([1,2,3],2)

#++++++++++++++++++++++++++++++
#INPUT: Obtain two words in a collection
#Split the two words and store them in two different collection objects
#Convert the words into lowercase
#Check to see if the first word includes the same letters as the first word
#Return true if it does
#Else return false
#OUTPUT: Return true if both words contain same letters. Else return false
def mutation(array)
  first_word = array[0].downcase
  second_word = array[1].downcase

  p first_word
  p second_word

  if first_word.include?(second_word)
      p true
  else
      p false
  end
end

mutation(["Mary", "Army"])

#Where do I belong
def where(array, num)
  sorted_array = array.sort
  num_to_place = num


  sorted_array.map do |number|
      if number > num_to_place
         index_num = sorted_array.index(number)
         index_num - 1
      end
  end
  #p sorted_array.index(highest_num)
end

where([40, 60, 30], 50)